### PR TITLE
ONNX improvements

### PR DESCRIPTION
**Pull request by ezyang**
```
commit 88a1e6acbc1d2aaf7731dadad69e0b4f8456cc29
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Tue Sep 19 09:03:08 2017 -0700

    Ask for the correct number of derivatives when tracing.
    
    - If you operate with TracingState, you MUST check if it is live.
      Otherwise you will segfault if it is expired; it is VALID for
      tracing states to become expired.
    
    - Tracing states can expire if they request backward tracing
      (which the tracer does by default).  We don't want this to
      happen for exports, which only look at forwards.  So make
      sure we set the correct num_derivatives.
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit 37941667d8e82ed1888c5b0aeb6b46a65875b89e
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Tue Sep 19 08:48:11 2017 -0700

    Make assertExpected work with Unicode strings in Python 2.
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit d0dc72d8cd9a0df7b094633a40486715680e3c60
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Mon Sep 18 18:55:33 2017 -0700

    More expect test improvements.
    
    - Print some diagnostic information when accepting new test output.
    
    - If it's the first time you ran an expect test, print out
      the output you got so it's easier to decide if you want
      to accept it.
    
    - Add infrastructure for expect-testing against exceptions
      (I'm going to use this in a later patch).
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit dd12823fade09e679629974ec3915c8585741d84
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Mon Sep 18 13:57:37 2017 -0700

    Tighten up the ONNX export interface
    
    - If a user accidentally attempts to export a model that is in training mode, the
      tracer may perturb the parameters (since modules like batchnorm will update
      their parameters.)  To prevent this from happening, we temporarily turn
      off training mode to make sure this doesn't happen.  Temporary is
      important, since model export should not actually affect the model
    
    - If you have a buggy model which is changing the parameters,
      it is much better for us to export the state_dict() *prior*
      to executing the model, because that is what we actually
      used as the inputs to the trace.  The state_dict() afterwards
      could be anything.
    
    - kwargs support never worked, so it's been excised.
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit 9e20d5b58b1d876f66effba438b096345bb0e687
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Tue Sep 19 14:29:17 2017 -0700

    Give better error message when symbolic() arguments to line up.
    
    Now we actually tell the user what operator was being translated
    when there was a failure.
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit 6d9b3455c6278f297ab479542bf9aa2a611d172b
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Tue Sep 19 11:08:14 2017 -0700

    Fix minor bug when --accept'ing commits.
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

commit 73d3972feeb577a40dbcf138e882769bbbd760f5
Author: Edward Z. Yang <ezyang@fb.com>
Date:   Tue Sep 19 10:53:57 2017 -0700

    Also fix AvgPool2d to follow new convention.
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>
```

CC @apaszke, @zdevito, @prigoyal, @houseroad

### CODE REVIEW

**Comment by ezyang** [comment]
@apaszke Take a look at the commit "Give better error message when symbolic() arguments to line up."

File: test/common.py [hunk 1]
```diff
@@ -296,20 +312,31 @@ def remove_prefix(text, prefix):
             expected_file += "-" + subname
         expected_file += ".expect"
         expected = None
+        try:
+            with open(expected_file) as f:
+                expected = f.read()
+        except IOError as e:
+            if e.errno == errno.ENOENT:
```
Comment (apaszke) [inline test/common.py]: This would be nicer (less indentation), but you don't need to fix it
```python
if e.errno != errno.ENOENT:
    raise
# continue code
```

File: test/common.py [hunk 2]
```diff
@@ -296,20 +312,31 @@ def remove_prefix(text, prefix):
...
+                    raise RuntimeError(
+                        ("I got this output for {}:\n\n{}\n\n"
+                         "No expect file exists; to accept the current output, run:\n"
+                         "python {} {} --accept").format(munged_id, s, __main__.__file__, munged_id))
+            else:
+                raise
         if ACCEPT:
-            with open(expected_file, 'w') as f:
-                f.write(s)
+            if expected == s:
```
Comment (apaszke) [inline test/common.py]: There's sth wrong with this code. Why do we have two places where we accept (overwrite) the file now? Opening for writing will create it anyway

File: test/test_onnx.py [hunk 3]
```diff
@@ -54,6 +55,16 @@ def test_transpose(self):
...
 
+    def test_concat(self):
+        class ConcatNet(nn.Module):
+            def forward(self, inputs):
+                return torch.cat(inputs, 1)
+
+        # volatile is of particular interest; there was a bug involving this
+        x = Variable(torch.randn(2, 3), volatile=True)
+        y = Variable(torch.randn(2, 3), volatile=True)
+        self.assertONNXExpected(export_to_string(ConcatNet(), ((x, y),)))
```
Comment (apaszke) [inline test/test_onnx.py]: no, no, no. Let's figure out a better strategy for testing onnx than dumping 200 expect files (for each op).

File: test/test_onnx.py [hunk 4]
```diff
@@ -76,11 +87,34 @@ def test_non_float_params(self):
...
+            @staticmethod
+            def symbolic(g, x):
+                assert False  # dead code
+
+            @staticmethod
+            def forward(ctx, x, y):
+                return x + y
+
+        class MyModel(Module):
+            def forward(self, x, y):
```
Comment (apaszke) [inline test/test_onnx.py]: We're doing so many of these `MyModel`s that we should probably have a helper that takes a lambda and wraps it in a module

File: test/test_onnx.py [hunk 5]
```diff
@@ -76,11 +87,34 @@ def test_non_float_params(self):
         torch._C._jit_pass_onnx(trace)
         self.assertONNXExpected(trace.export(initializers))
 
+    def test_symbolic_mismatch(self):
+        class MyFun(Function):
+            @staticmethod
+            def symbolic(g, x):
+                assert False  # dead code
```
Comment (apaszke) [inline test/test_onnx.py]: doesn't look dead to me? either explain it or remove the comment

File: torch/csrc/jit/passes/onnx.cpp [hunk 6]
```diff
@@ -26,6 +26,12 @@ bool hasUsedHandle(Node *node) {
 
 // Transform PythonOps and Cpp Ops into Node's that match ONNX semantics.
 void ToONNX(std::shared_ptr<tracer::TracingState>& state) {
+  // Check that the tracing state is live (it should be, because
+  // you were supposed to request zero derivatives.)
+  if (state->is_expired()) {
+    throw std::runtime_error("Tracing state is expired!  You should run the tracer with num_derivatives=0");
```
Comment (apaszke) [inline torch/csrc/jit/passes/onnx.cpp]: No, `num_derivatives=0` is not specific to `ToONNX` pass in any way. We use it e.g. when running `word_language_model` example with `num_derivatives=1`

File: torch/csrc/jit/passes/onnx.cpp [hunk 7]
```diff
@@ -129,9 +141,10 @@ void ToONNX(std::shared_ptr<tracer::TracingState>& state) {
       py_symbolic_args[input_nr++] = obj;
     }
     // Call the symbolic function
-    py::object raw_output = py::reinterpret_steal<py::object>(
-        PyObject_CallObject(pyobj.attr("symbolic").ptr(), py_symbolic_args.ptr()));
-    if (!raw_output) throw py::error_already_set();
+    py::object onnx = py::module::import("torch.onnx");
+    // Use a little trampoline function so we can give good error messages
+    // upon argument mismatch
+    py::object raw_output = onnx.attr("run_symbolic")(op->name(), pyobj.attr("symbolic"), py_symbolic_args);
```
Comment (apaszke) [inline torch/csrc/jit/passes/onnx.cpp]: This looks terribly slow, and I really can't understand why do we need a Python trampoline just to append "occurred when translating X". You can easily do that in 3 lines of C, and avoid passing control around 3 files

File: torch/onnx.py [hunk 8]
```diff
@@ -23,35 +44,60 @@ def export(model, args, f, export_params=True, kwargs=None, verbose=False):
...
+        training (bool, default False): export the model in training mode.  At
+            the moment, ONNX is oriented towards exporting models for inference
+            only, so you will generally not need to set this to True.
     """
-    _export(model, args, f, export_params, kwargs, verbose)
+    _export(model, args, f, export_params, verbose, training)
 
 
-def _export(model, args, f, export_params=True, kwargs=None, verbose=False):
+def _export(model, args, f, export_params=True, verbose=False, training=False):
```
Comment (apaszke) [inline torch/onnx.py]: why did you remove `kwargs`?

File: torch/onnx.py [hunk 9]
```diff
@@ -23,35 +44,60 @@ def export(model, args, f, export_params=True, kwargs=None, verbose=False):
...
+    # here.  This shouldn't happen, because we've turned off training
+    # mode and networks in inference mode should be purely functional.
+    # But it's user code; anything can happen!
+    params = None
+    if export_params:
+        # NB: OrderedDict values is not actually a list, but trace.export is
+        # not duck-typed and expects an actual list.
+        params = list(model.state_dict().values())
+    # It's important to run the model in inference mode when exporting;
+    # otherwise internal buffers may get updated, dropout gets applied, etc.
```
Comment (apaszke) [inline torch/onnx.py]: this code doesn't hard-code inference mode. It's only `False` by default, but can be overriden

File: torch/onnx.py [hunk 10]
```diff
@@ -23,35 +44,60 @@ def export(model, args, f, export_params=True, kwargs=None, verbose=False):
...
     # Special case for common case of passing a single Variable
     if isinstance(args, torch.autograd.Variable):
         args = (args, )
-    if not kwargs:
-        kwargs = {}
-    trace, torch_out = torch.jit.record_trace(model, *args, **kwargs)
+    # Look at the state_dict *prior* to running the model, as this
+    # accurately captures what inputs we actually passed to the model.
+    # If we run it afterwards, a buggy forward pass could have
+    # added/deleted parameters, changing the structure of the state_dict.
```
Comment (apaszke) [inline torch/onnx.py]: If you're worried about this, then check it, instead of cloning the thing beforehand and pretending it's ok

File: torch/onnx.py (hunk 8 above)
Comment (ezyang) [reply torch/onnx.py]: They didn't actually do the right thing. The docs implied that they would be passed to the model as keyword arguments, but they didn't actually do that.

File: torch/csrc/jit/passes/onnx.cpp (hunk 7 above)
Comment (ezyang) [reply torch/csrc/jit/passes/onnx.cpp]: Invoking symbolic functions shouldn't ever be the bottleneck for export. But fine, I'll do it in C. (One other motivation for having a Python trampoline is to ease up debugging: if you want to see what symbolic functions are being called, there's now one place you can edit without having to recompile the C++)

File: test/common.py (hunk 2 above)
Comment (ezyang) [reply test/common.py]: I'll factor the duplicate code into a helper method

File: test/test_onnx.py (hunk 3 above)
Comment (ezyang) [reply test/test_onnx.py]: Can we discuss this in today's standup (i.e., does this block the patch?) I think there are benefits to dumping expect files, at least for some of them.

For this particular patch, I added this test because concat export was segfaulting. So, if it was just the segfault I wanted to test, I could get rid of the assertONNXExpected. But let's talk about it.

File: torch/csrc/jit/passes/onnx.cpp (hunk 6 above)
Comment (ezyang) [reply torch/csrc/jit/passes/onnx.cpp]: OK, I'm going to do something like:

```
diff --git a/torch/csrc/jit/passes/onnx.cpp b/torch/csrc/jit/passes/onnx.cpp
index 842d79f..0596d05 100644
--- a/torch/csrc/jit/passes/onnx.cpp
+++ b/torch/csrc/jit/passes/onnx.cpp
@@ -29,7 +29,7 @@ void ToONNX(std::shared_ptr<tracer::TracingState>& state) {
   // Check that the tracing state is live (it should be, because
   // you were supposed to request zero derivatives.)
   if (state->is_expired()) {
-    throw std::runtime_error("Tracing state is expired!  You should run the tracer with num_derivatives=0");
+    throw std::logic_error("ToONNX: tracing state is expired");
   }
 
   auto new_graph = std::make_shared<Graph>();
```

File: torch/onnx.py (hunk 10 above)
Comment (ezyang) [reply torch/onnx.py]: Well, technically, no matter how the user changes the parameters list after the fact, the state_dict prior to running the trace *will be accurate* for the trace.

Comment (ezyang) [reply torch/onnx.py]: I will check that the keys match.

File: torch/csrc/jit/passes/onnx.cpp (hunk 7 above)
Comment (ezyang) [reply torch/csrc/jit/passes/onnx.cpp]: Yeah, uh.... I have no idea how to implement this in three lines of C (I need string concatenation! String allocation! Tuple allocation! Attribute setting! Reraising of exceptions! And handle the lifetimes of all the Python objects correctly!)

Comment (ezyang) [reply torch/csrc/jit/passes/onnx.cpp]: For posterity, I spent an hour trying to make this work out, and then gave up in the end.

**Comment by ezyang** [comment]
Ready to merge :)
